jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java

```diff
@@ -67,68 +69,154 @@ public ArrayByteBufferPool(int minCapacity, int factor, int maxCapacity)
         _direct = new ByteBufferPool.Bucket[length];
         _indirect = new ByteBufferPool.Bucket[length];
-
-        int capacity = 0;
-        for (int i = 0; i < _direct.length; ++i)
-        {
-            capacity += _factor;
-            _direct[i] = new ByteBufferPool.Bucket(this, capacity, maxQueueLength);
-            _indirect[i] = new ByteBufferPool.Bucket(this, capacity, maxQueueLength);
-        }
     }
 
     @Override
     public ByteBuffer acquire(int size, boolean direct)
     {
-        ByteBufferPool.Bucket bucket = bucketFor(size, direct);
+        int capacity = size < _minCapacity ? size : (bucketFor(size) + 1) * getCapacityFactor();
+        ByteBufferPool.Bucket bucket = bucketFor(size, direct, null);
         if (bucket == null)
-            return newByteBuffer(size, direct);
-        return bucket.acquire(direct);
+            return newByteBuffer(capacity, direct);
+        ByteBuffer buffer = bucket.acquire();
+        if (buffer == null)
+            return newByteBuffer(capacity, direct);
+        decrementMemory(buffer);
+        return buffer;
     }
 
     @Override
     public void release(ByteBuffer buffer)
     {
-        if (buffer != null)
+        if (buffer == null)
+            return;
+        boolean direct = buffer.isDirect();
+        ByteBufferPool.Bucket bucket = bucketFor(buffer.capacity(), direct, this::newBucket);
+        if (bucket != null)
         {
-            ByteBufferPool.Bucket bucket = bucketFor(buffer.capacity(), buffer.isDirect());
-            if (bucket != null)
-                bucket.release(buffer);
+            bucket.release(buffer);
+            incrementMemory(buffer);
+            releaseExcessMemory(direct, this::clearOldestBucket);
         }
     }
 
-    @ManagedOperation(value = "Clears this ByteBufferPool", impact = "ACTION")
+    private Bucket newBucket(int key)
+    {
+        return new Bucket(this, key * getCapacityFactor(), getMaxQueueLength());
+    }
+
+    @Override
     public void clear()
     {
+        super.clear();
         for (int i = 0; i < _direct.length; ++i)
         {
-            _direct[i].clear();
-            _indirect[i].clear();
+            Bucket bucket = _direct[i];
+            if (bucket != null)
+                bucket.clear();
+            _direct[i] = null;
+            bucket = _indirect[i];
+            if (bucket != null)
+                bucket.clear();
+            _indirect[i] = null;
         }
     }
 
-    private ByteBufferPool.Bucket bucketFor(int capacity, boolean direct)
+    private void clearOldestBucket(boolean direct)
     {
-        if (capacity <= _minCapacity)
+        long oldest = 0;
+        int index = -1;
+        Bucket[] buckets = bucketsFor(direct);
+        long now = System.nanoTime();
+        for (int i = 0; i < buckets.length; ++i)
+        {
+            Bucket bucket = buckets[i];
+            if (bucket == null)
+                continue;
+            long age = now - bucket.getLastUpdate();
```

Comment:
Do we actually need to calculate age?  Can't we just look for the smallest lastUpdate?

Reply:
`System.nanoTime()` can return a negative value; while `oldest` can be set to `Long.MIN_VALUE`, it feels safer to me to calculate a difference that will always be positive.

Reply:
Actually `Long.MAX_VALUE`. I changed my mind and the code to use that.